src/c++/perf_analyzer/client_backend/openai/http_client.cc

```diff
@@ -0,0 +1,267 @@
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "http_client.h"
+
+#include <functional>
+#include <iostream>
+
+namespace triton { namespace perfanalyzer { namespace clientbackend {
+namespace openai {
+
+HttpRequest::HttpRequest(
+    std::function<void(HttpRequest*)>&& completion_callback, const bool verbose)
+    : completion_callback_(std::move(completion_callback)), verbose_(verbose)
+{
+}
+
+HttpRequest::~HttpRequest()
+{
+  if (header_list_ != nullptr) {
+    curl_slist_free_all(header_list_);
+    header_list_ = nullptr;
+  }
+}
+
+void
+HttpRequest::AddInput(uint8_t* buf, size_t byte_size)
+{
+  data_buffers_.push_back(std::pair<uint8_t*, size_t>(buf, byte_size));
+  total_input_byte_size_ += byte_size;
+}
+
+void
+HttpRequest::GetNextInput(uint8_t* buf, size_t size, size_t* input_bytes)
+{
+  *input_bytes = 0;
+
+  while (!data_buffers_.empty() && size > 0) {
+    const size_t csz = std::min(data_buffers_.front().second, size);
+    if (csz > 0) {
+      const uint8_t* input_ptr = data_buffers_.front().first;
+      std::copy(input_ptr, input_ptr + csz, buf);
+      size -= csz;
+      buf += csz;
+      *input_bytes += csz;
+
+      data_buffers_.front().first += csz;
+      data_buffers_.front().second -= csz;
+    }
+    if (data_buffers_.front().second == 0) {
+      data_buffers_.pop_front();
+    }
+  }
+}
+
+HttpClient::HttpClient(
+    const std::string& server_url, bool verbose,
+    const HttpSslOptions& ssl_options)
+    : url_(server_url), verbose_(verbose), ssl_options_(ssl_options)
+{
+  auto* ver = curl_version_info(CURLVERSION_NOW);
+  if (ver->features & CURL_VERSION_THREADSAFE == 0) {
```

Comment:
Can't compile because of this line.

`CURL_VERSION_THREADSAFE` was added in libcurl 7.84.0 (https://curl.se/libcurl/c/curl_version_info.html#threadsafe).

PA builds with libcurl 7.81.0 because that's what gets installed by default from apt in ubuntu 22.04:
```
$ docker run -it ubuntu:22.04 bash -c 'apt update ; apt list -a libcurl4-openssl-dev'
...
libcurl4-openssl-dev/jammy-updates,jammy-security 7.81.0-1ubuntu1.15 amd64
libcurl4-openssl-dev/jammy 7.81.0-1 amd64

Reply:
@GuanLuo

Reply:
Does it not use curl from third party repo?

Reply:
If that is the case, we can add a class static mutex and ensure the `curl_global_init` and `curl_global_cleanup` are protected by the singleton mutex

Reply:
Where in the cmake would it indicate that it's compiling with the curl from the third party repo?

I see this:
https://github.com/triton-inference-server/client/blob/e581c8178d4209db9d2b3f77515f1e8bb4b2811d/CMakeLists.txt#L158

But `CURL_DIR` doesn't seem to be used anywhere else.

There's a bunch of places in cmake that call `find_package(CURL ...)`, but I assume that would find nothing without `apt install libcurl4-openssl-dev`.

Reply:
Might be better to use a feature available in 7.81.0 to achieve the same ends, for now.

Reply:
I believe `CURL_DIR` is hint for CMake find_package that on `find_package(XXX ...)`, it checks if `XXX_DIR` has been defined.

Reply:
Okay, so if the non-thread safe version is used, then the global init/cleanup must even be called when there is only one thread in the process to be completely safe..
> If the global constant functions are not thread safe, then the basic rule for constructing a program that uses libcurl is this: Call [curl_global_init](https://curl.se/libcurl/c/curl_global_init.html), with a CURL_GLOBAL_ALL argument, immediately after the program starts, while it is still only one thread and before it uses libcurl at all. Call [curl_global_cleanup](https://curl.se/libcurl/c/curl_global_cleanup.html) immediately before the program exits, when the program is again only one thread and after its last use of libcurl.

Reply:
I merged in Guan's changes into this branch. Are there any more opens or concerns from this thread?
https://github.com/triton-inference-server/client/pull/476/commits/fd6f2af735b45c5f5352da2bdc83e3486071fe5b